src/service/action-impl.js

```diff
@@ -381,33 +362,39 @@ export class ActionService {
       return;
     }
 
-    for (let i = 0; i < action.actionInfos.length; i++) {
-      const actionInfo = action.actionInfos[i];
+    // Invoke actions serially, where each action waits for its predecessor
+    // to complete. `currentPromise` is the i'th promise in the chain.
+    let currentPromise = null;
 
-      // Replace any expressions in args with data in `event`.
+    action.actionInfos.forEach((actionInfo, i) => {
+      // Replace any variables in args with data in `event`.
       const args = dereferenceExprsInArgs(actionInfo.args, event);
 
-      // Global target, e.g. `AMP`.
-      const globalTarget = this.globalTargets_[actionInfo.target];
-      if (globalTarget) {
-        const invocation = new ActionInvocation(
-            this.root_,
-            actionInfo.method,
-            args,
-            action.node,
-            event,
-            trust);
-        globalTarget(invocation);
-      } else {
+      const invoke = () => {
+        // Global target, e.g. `AMP`.
+        const globalTarget = this.globalTargets_[actionInfo.target];
+        if (globalTarget) {
+          const invocation = new ActionInvocation(this.root_, actionInfo.method,
+              args, action.node, event, trust);
+          return globalTarget(invocation, i, action.actionInfos);
+        }
+
+        // Element target via `id` attribute.
         const target = this.root_.getElementById(actionInfo.target);
         if (target) {
-          this.invoke_(target, actionInfo.method, args,
-              action.node, event, trust, actionInfo);
+          const invocation = new ActionInvocation(target, actionInfo.method,
+              args, action.node, event, trust);
+          return this.invoke_(invocation, actionInfo);
         } else {
           this.actionInfoError_('target not found', actionInfo, target);
         }
-      }
-    }
+      };
+
+      // Wait for the previous action, if applicable.
+      currentPromise = (currentPromise)
+          ? currentPromise.then(() => invoke())
+          : invoke();
```

Comment:
If `invoke()` returns null, then during the next iteration it'll fall into the "else" branch of the ternary again, no? Note that `currentPromise` is initialized to `null`.